stream.go

```diff
@@ -0,0 +1,177 @@
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"io"
+)
+
+type DecryptReader struct {
+	r          *bufio.Reader
+	mode       cipher.BlockMode
+	cg         CredsGenerator
+	passphrase []byte
+}
+
+func NewReader(passphrase string, r io.Reader, cg CredsGenerator) *DecryptReader {
+	return &DecryptReader{
+		r:          bufio.NewReader(r),
+		cg:         cg,
+		passphrase: []byte(passphrase),
+	}
+}
+
+func (d *DecryptReader) Init() error {
+	if d.mode != nil {
+		return nil
+	}
+
+	saltHeader := make([]byte, aes.BlockSize)
+	if _, err := io.ReadFull(d.r, saltHeader); err != nil {
+		return fmt.Errorf("read salt header failed: %w", err)
+	}
+
+	if string(saltHeader[:8]) != "Salted__" {
+		return fmt.Errorf("does not appear to have been encrypted with OpenSSL, salt header missing")
+	}
+
+	salt := saltHeader[8:]
+
+	creds, err := d.cg(d.passphrase, salt)
+	if err != nil {
+		return err
+	}
+
+	c, err := aes.NewCipher(creds.Key)
+	if err != nil {
+		return err
+	}
+	d.mode = cipher.NewCBCDecrypter(c, creds.IV)
+	return nil
+}
+
+func (d *DecryptReader) Read(b []byte) (int, error) {
+	if d.mode == nil {
+		if err := d.Init(); err != nil {
+			return 0, fmt.Errorf("init failed: %w", err)
+		}
+	}
+
+	size := (len(b) / aes.BlockSize) * aes.BlockSize
+
+	// log.Println(size, len(b))
+
+	if size == 0 {
+		return 0, nil
+	}
+
+	n, err := d.r.Read(b[:size])
+	if err != nil {
+		return 0, err
+	}
+
+	d.mode.CryptBlocks(b[:n], b[:n])
+
+	if _, err = d.r.Peek(aes.BlockSize); errors.Is(err, io.EOF) {
+		n = n - int(b[n-1])
+	}
+
+	return n, nil
+}
+
+type EncryptWriter struct {
+	mode       cipher.BlockMode
+	w          io.Writer
+	cg         CredsGenerator
+	passphrase []byte
+	buf        []byte
+}
+
+func NewWriter(passphrase string, w io.Writer, cg CredsGenerator) *EncryptWriter {
+	return &EncryptWriter{
+		w:          w,
+		cg:         cg,
+		passphrase: []byte(passphrase),
+	}
+}
+
+func (e *EncryptWriter) Init() error {
+	if e.mode != nil {
+		return nil
+	}
+
+	salt := make([]byte, 8) // Generate an 8 byte salt
+	_, err := io.ReadFull(rand.Reader, salt)
+	if err != nil {
+		return err
+	}
+
+	_, err = e.w.Write(append([]byte("Salted__"), salt...))
+	if err != nil {
+		return err
+	}
+
+	creds, err := e.cg([]byte(e.passphrase), salt)
+	if err != nil {
+		return err
+	}
+
+	c, err := aes.NewCipher(creds.Key)
+	if err != nil {
+		return err
+	}
+	e.mode = cipher.NewCBCEncrypter(c, creds.IV)
+	return nil
+}
+
+func (e *EncryptWriter) Write(b []byte) (int, error) {
+	if e.mode == nil {
+		if err := e.Init(); err != nil {
+			return 0, err
+		}
+	}
+
+	originSize := len(b)
+
+	if e.buf != nil {
+		b = append(e.buf, b...)
+		e.buf = nil
+	}
+
+	size := (len(b) / aes.BlockSize) * aes.BlockSize
+
+	if remain := len(b) - size; remain > 0 {
+		e.buf = b[size:]
+	}
+
+	if size == 0 {
+		return originSize, nil
+	}
+
+	e.mode.CryptBlocks(b[:size], b[:size])
+
+	_, err := e.w.Write(b[:size])
+
+	return originSize, err
+}
+
+func (e *EncryptWriter) Flush() error {
```

Comment:
When adding the doc-comment for this please stress this **MUST** be called in order to have all data written.

Also: Would it maybe make sense to implement a `io.ReadCloser` and do this inside the `Close() error` function? I think that might be more commonly used in the Go world…